Group counts 2

nicpottier opened the pull request:
This is part 2, which adds the fun of the triggers and switches to using system groups for All Contacts, Failed Contacts etc..

ericnewcomer commented on temba/contacts/migrations/0013_contactgroup_group_type.py:
```diff
@@ -0,0 +1,31 @@
+            field=models.CharField(default='U', help_text='What type of group it is, either user defined or one of our system groups', max_length=1, choices=[('A', 'All Contacts'), ('B', 'Blocked Contacts'), ('F', 'Failed Contacts'), ('U', 'User Defined Groups')]),
+        ),
+
+    ]
```
Can I just say I love 1.7 migrations?

ericnewcomer commented on temba/contacts/migrations/0014_populate_system_groups.py:
```diff
@@ -0,0 +1,52 @@
+        Contact = apps.get_model('contacts', 'Contact')
+
+        print "Creating group.."
+
```
Nice to avoid prints in migrations that will print on empty dbs (ie, tests)

nicpottier commented on temba/contacts/models.py:
```diff
@@ -743,27 +752,41 @@ def block(self):
+            group.update_contacts([self], False)
 
-    def unblock(self):
+        ContactGroup.system_groups.get(org=self.org, group_type=BLOCKED_CONTACTS_GROUP).contacts.add(self)
```
Another way to do this would be to create triggers that add/remove from the system groups when states change, but I'm trying to restrain myself to only using triggers for counts for the time being.

nicpottier commented on temba/orgs/models.py:
```diff
@@ -895,8 +877,21 @@ def has_nexmo_number(self):
+        self.groups.create(name='Blocked Contacts', group_type=BLOCKED_CONTACTS_GROUP,
+                           created_by=self.created_by, modified_by=self.modified_by)
+        self.groups.create(name='Failed Contacts', group_type=FAILED_CONTACTS_GROUP,
+                           created_by=self.created_by, modified_by=self.modified_by)
```
Again, maybe we should just put this in a trigger based on Org creation? That would always guarantee that no matter what these groups exist.. 

Is that getting too crazy?

ericnewcomer replied:
The issue with moving things to triggers is it creates a certain amount of magic in the system, especially down the line when this behavior changes, it'll be easy to miss. I'd prefer this logic stay in the code and save triggers primarily for performance considerations.

nicpottier commented:
@rowanseymour Appreciate your eyes on this tomorrow, especially in light of me breaking mage last time..

nicpottier commented:
@ericnewcomer Not sure where your comment went, but ya, I'd agree with that. Currently the "System Groups" are all updated (contacts added and removed in code) so I suppose their creation being in code makes sense as well.

If we switched to the contacts in those groups being updated by a trigger (say when is_blocked is changed) then I would say the argument for the creation of those groups being a trigger would probably be stronger, but the current balance is probably the right one for the time being.

ericnewcomer commented on temba/contacts/models.py:
```diff
@@ -1231,20 +1256,52 @@ class Meta:
     @classmethod
     def get_or_create(cls, org, user, name):
-        existing = ContactGroup.objects.filter(name=name, org=org, is_active=True).first()
+        existing = ContactGroup.user_groups.filter(name=name, org=org, is_active=True).first()
```
Group name lookup should be case insensitive

rowanseymour commented on temba/contacts/models.py:
```diff
@@ -743,27 +752,41 @@ def block(self):
-    def unblock(self):
+        ContactGroup.system_groups.get(org=self.org, group_type=BLOCKED_CONTACTS_GROUP).contacts.add(self)
+
+    def restore(self):
```
Minor point but I think `unblock` was clearer than `restore` since that could also mean un-deleting a contact

nicpottier replied:
I actually prefer unblock as well, but we had this undefined line in the call stack where we decided that restore is now unblock instead and that felt super weird. (because the Action stuff is trying to reuse functionality etc..)

rowanseymour commented on temba/contacts/models.py:
```diff
@@ -179,6 +179,13 @@ def anon_identifier(self):
         return "%010d" % self.id
 
+    @property
+    def user_groups(self):
```
`custom_groups` might be clearer by avoiding confusion with actual user groups from the `auth` app

rowanseymour commented on temba/contacts/models.py:
```diff
@@ -437,9 +444,13 @@ def get_or_create(cls, org, user, name=None, urns=None, incoming_channel=None, u
                 contact = Contact.objects.create(**updated_attrs)
 
+                # add it to our All Contacts group
+                if not contact.is_test:
```
Need to add this logic to Mage as well.... which again makes me wish we weren't replicating Temba functionality in Mage and instead only using to push Twitter events

nicpottier replied:
I don't think so, this is actually handled by the "is_new_contact" path in mage.py I believe.

rowanseymour replied:
Ah ok that works
